lib/auth/join_azure.go

```diff
@@ -0,0 +1,371 @@
+	Encoding  string `json:"encoding"`
+	Signature string `json:"signature"`
+}
+
+type plan struct {
+	Name      string `json:"name"`
+	Product   string `json:"product"`
+	Publisher string `json:"publisher"`
+}
+
+type timestamp struct {
+	CreatedOn string `json:"createdOn"`
+	ExpiresOn string `json:"expiresOn"`
+}
+
+type attestedData struct {
+	LicenseType    string    `json:"licenseType"`
+	Nonce          string    `json:"nonce"`
+	Plan           plan      `json:"plan"`
+	Timestamp      timestamp `json:"timestamp"`
+	ID             string    `json:"vmId"`
+	SubscriptionID string    `json:"subscriptionId"`
+	SKU            string    `json:"sku"`
+}
+
+type accessTokenClaims struct {
+	jwt.Claims
+	ResourceID string `json:"xms_mirid"`
+	TenantID   string `json:"tid"`
+	Version    string `json:"ver"`
+}
+
+type azureVerifyTokenFunc func(ctx context.Context, rawIDToken string) (*accessTokenClaims, error)
+
+type azureRegisterConfig struct {
+	clock    clockwork.Clock
+	certs    []*x509.Certificate
+	verify   azureVerifyTokenFunc
+	vmClient azure.VirtualMachinesClient
+}
+
+func azureVerifyFuncFromOIDCVerifier(cfg *oidc.Config) azureVerifyTokenFunc {
+	return func(ctx context.Context, rawIDToken string) (*accessTokenClaims, error) {
+		token, err := jwt.ParseSigned(rawIDToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		var unverifiedClaims accessTokenClaims
+		if err := token.UnsafeClaimsWithoutVerification(&unverifiedClaims); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		provider, err := oidc.NewProvider(ctx, fmt.Sprintf("https://sts.windows.net/%v/", unverifiedClaims.TenantID))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		verifiedToken, err := provider.Verifier(cfg).Verify(ctx, rawIDToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		var tokenClaims accessTokenClaims
+		if err := verifiedToken.Claims(&tokenClaims); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return &tokenClaims, nil
+	}
+}
+
+func (cfg *azureRegisterConfig) CheckAndSetDefaults(ctx context.Context) error {
+	if cfg.clock == nil {
+		cfg.clock = clockwork.NewRealClock()
+	}
+	if cfg.verify == nil {
+		oidcConfig := &oidc.Config{
+			ClientID: azureAccessTokenAudience,
+			Now:      cfg.clock.Now,
+		}
+		cfg.verify = azureVerifyFuncFromOIDCVerifier(oidcConfig)
+	}
+
+	if cfg.certs == nil {
+		certs, err := getAzureCerts()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		cfg.certs = certs
+	}
+	return nil
+}
+
+type azureRegisterOption func(cfg *azureRegisterConfig)
+
+func withCerts(certs []*x509.Certificate) azureRegisterOption {
+	return func(cfg *azureRegisterConfig) {
+		cfg.certs = certs
+	}
+}
+
+func withVerifyFunc(verify azureVerifyTokenFunc) azureRegisterOption {
+	return func(cfg *azureRegisterConfig) {
+		cfg.verify = verify
+	}
+}
+
+func withVMClient(vmClient azure.VirtualMachinesClient) azureRegisterOption {
+	return func(cfg *azureRegisterConfig) {
+		cfg.vmClient = vmClient
+	}
+}
```

Comment:
I notice these are only being used in a test right now. Should they be being used elsewhere, or be moved to the test file?